nexus/db-queries/src/transaction_retry.rs

```diff
@@ -0,0 +1,141 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Helper types for performing automatic transaction retries
+
+use chrono::Utc;
+use oximeter::{types::Sample, Metric, MetricsError, Target};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+// Identifies "which" transaction is retrying
+#[derive(Debug, Clone, Target)]
+struct Transaction {
+    name: String,
+}
+
+// Identifies that a retry has occurred, and track how long
+// the transaction took (either since starting, or since the last
+// retry failure was recorded).
+#[derive(Debug, Clone, Metric)]
+struct RetryData {
+    #[datum]
+    latency: f64,
+}
+
+// Collects all transaction retry samples
+#[derive(Debug, Default, Clone)]
+pub(crate) struct Producer {
+    samples: Arc<Mutex<Vec<Sample>>>,
+}
+
+impl Producer {
+    pub(crate) fn new() -> Self {
+        Self { samples: Arc::new(Mutex::new(vec![])) }
+    }
+
+    fn append(
+        &self,
+        transaction: &Transaction,
+        latency: Duration,
+    ) -> Result<(), MetricsError> {
+        let sample = Sample::new_with_timestamp(
+            Utc::now(),
+            transaction,
+            &RetryData { latency: latency.as_secs_f64() },
+        )?;
+        self.samples.lock().unwrap().push(sample);
+        Ok(())
+    }
+}
+
+struct RetryHelperInner {
+    start: chrono::DateTime<Utc>,
+    attempts: usize,
+}
+
+impl RetryHelperInner {
+    fn new() -> Self {
+        Self { start: Utc::now(), attempts: 1 }
+    }
+
+    fn tick(&mut self) -> Self {
+        let start = self.start;
+        let attempts = self.attempts;
+
+        self.start = Utc::now();
```

Comment:
Resetting `start` ensures that each `RetryData` metric shows up separately, rather than each one appearing to take longer and longer.